states/pong2.mod.py

```diff
@@ -1,313 +1,253 @@
+                pixels = []
+                for _ in range(self.winh):
+                    pixels.append([])
+                    for _ in range(self.winw):
+                        pixels[-1].append([0, 0, 0])
+                # draw all elements
+                ball = self.game.ball
+                ball_x = min(max(0, ball.x), self.game.dims["width"] - 1)
+                ball_y = min(max(0, ball.y), self.game.dims["height"] - 1)
+                pixels[round(ball_y)][round(ball_x)] = WHITE
+                for player in [self.game.p1, self.game.p2]:
+                    for y in range(player.y, player.y + player.height):
+                        pixels[y][player.x] = player.color()
                 for y in range(0, self.winh):
-                    for x in range(0, self.winw):
-                        frame += get_pixel(x, y, p1_win, p2_win)
-
-                for i in range(150):
-                    self.output_frame(frame)
-                    sleep(self.frame_delay)
-
-            sleep(self.frame_delay)
-
-    def receive(self, context):
-        HOST = '0.0.0.0'  # Standard loopback interface address (localhost)
-        PORT = 9999  # Port to listen on (non-privileged ports are > 1023)
-
+                    if y % 2 == 0:
+                        pixels[y][int(self.winw / 2)] = WHITE
+                scores = [
+                    {
+                        "score": list(str(self.game.p1.score)),
+                        "color": GREEN if self.game.p1.has_won else WHITE,
+                        "center": round(self.game.dims["width"] / 4 - 1)
+                    },
+                    {
+                        "score": list(str(self.game.p2.score)),
+                        "color": GREEN if self.game.p2.has_won else WHITE,
+                        "center": round((self.game.dims["width"] / 4 * 3) - 1)
+                    }
+                ]
+                for sc in scores:
+                    digit_left = len(sc["score"]) * 2
+                    for index, digit in enumerate(sc["score"]):
+                        for y, row in enumerate(number(digit)):
+                            for x, pixel in enumerate(row):
+                                if pixel:
+                                    pos = sc["center"] - digit_left + x
+                                    pos += index * 4
+                                    pixels[y + 1][pos] = sc["color"]
+                # flatten, convert and write buffer to display
+                self.output_frame(bytes(flatten(pixels)))
+                sleep(1 / 30)
+
+    def receive(self):
         with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
             s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
             s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
-            s.bind((HOST, PORT))
+            s.bind(("0.0.0.0", 9999))
             s.listen()
-            while not context['killed']:
+            while True:
```

Comment:
`while not self.killed`

Anders krijgen we listener leaks

Reply:
maar we willen altijd luisteren naar request om een game te kunnen starten toch?

Reply:
de init() van een state wordt bij elk 'gebruik' van de state opnieuw opgeroepen. Sterker nog: een instance van State wordt maar 1 keer gebruikt en daarna nooit meer (want Threads kun je geen 2 keer run()-en)